ec/src/models/mnt6/mod.rs

```diff
@@ -112,27 +112,36 @@ impl<P: MNT6Parameters> MNT6<P> {
 
         // code below gets executed for all bits (EXCEPT the MSB itself) of
         // mnt6_param_p (skipping leading zeros) in MSB to LSB order
-        for (bit, dc) in BitIteratorBE::without_leading_zeros(P::ATE_LOOP_COUNT)
-            .skip(1)
-            .zip(&q.double_coefficients)
-        {
+        let y_over_twist_neg = -q.y_over_twist;
+        for (bit, dc) in P::ATE_LOOP_COUNT.iter().skip(1).zip(&q.double_coefficients) {
             let g_rr_at_p = Fp6::new(
                 dc.c_l - &dc.c_4c - &(dc.c_j * &p.x_twist),
                 dc.c_h * &p.y_twist,
             );
 
             f = f.square() * &g_rr_at_p;
 
-            if bit {
+            // Compute l_{R,Q}(P) if bit == 1, and l_{R,-Q}(P) if bit == -1
+            let g_rq_at_p;
+            if *bit == 1 {
                 let ac = &q.addition_coefficients[add_idx];
                 add_idx += 1;
 
-                let g_rq_at_p = Fp6::new(
+                g_rq_at_p = Fp6::new(
                     ac.c_rz * &p.y_twist,
                     -(q.y_over_twist * &ac.c_rz + &(l1_coeff * &ac.c_l1)),
                 );
-                f *= &g_rq_at_p;
+            } else if *bit == -1 {
+                let ac = &q.addition_coefficients[add_idx];
+                add_idx += 1;
+                g_rq_at_p = Fp6::new(
+                    ac.c_rz * &p.y_twist,
+                    -(y_over_twist_neg * &ac.c_rz + &(l1_coeff * &ac.c_l1)),
+                );
+            } else {
+                continue;
             }
```

Comment:
```suggestion
            } else if *bit == 0 {
                continue;
            } else {
                unimplemented!();
            }
```

Reply:
It should be `unreachable!()`, not `unimplemented!()`; we can only ever hit the +1 or -1 cases.

Reply:
Unless the developer makes a mistake?

Reply:
But I got it---it is not unimplemented but it is something the program should panic. Let me change that.